storage/remote/queue_manager.go

```diff
@@ -302,11 +341,96 @@ func NewQueueManager(reg prometheus.Registerer, metrics *queueManagerMetrics, lo
 	}
 
 	t.watcher = wal.NewWatcher(reg, wal.NewWatcherMetrics(reg), logger, client.Name(), t, walDir)
+	if t.cfg.SendMetadata {
+		t.metadataWatcher = NewMetadataWatcher(logger, sm, client.Name(), t, t.cfg.MetadataSendInterval, flushDeadline)
+	}
 	t.shards = t.newShards()
 
 	return t
 }
 
+// AppendMetadata queues metadata to be sent to the remote storage. Metadata is sent all at once and is not parallelized.
+func (t *QueueManager) AppendMetadata(ctx context.Context, metadata []scrape.MetricMetadata) {
+	metadataSet := map[scrape.MetricMetadata]struct{}{}
+
+	// We create a set of the metadata by deduplicating based on the attributes of a
+	// scrape.MetricMetadata. In this case, a combination of metric name, help, type, and unit.
+	for _, entry := range metadata {
+		metadataSet[entry] = struct{}{}
+	}
+
+	mm := make([]prompb.MetricMetadata, 0, len(metadataSet))
+	for entry := range metadataSet {
+		mm = append(mm, prompb.MetricMetadata{
+			MetricName: entry.Metric,
+			Help:       entry.Help,
+			Type:       metricTypeToMetricTypeProto(entry.Type),
+			Unit:       entry.Unit,
+		})
+	}
+
+	err := t.sendMetadataWithBackoff(ctx, mm)
+
+	if err != nil {
+		level.Error(t.logger).Log("msg", "non-recoverable error while sending metadata", "count", len(metadata), "err", err)
+	}
+}
+
+func (t *QueueManager) sendWriteRequestWithBackoff(ctx context.Context, req []byte, reportFailureFunc func()) (elapsed float64, err error) {
+	backoff := t.cfg.MinBackoff
+
+	for {
+		select {
+		case <-ctx.Done():
+			return -1, ctx.Err()
+		default:
+		}
+
+		begin := time.Now()
+		err := t.client.Store(context.Background(), req)
+		elapsed = time.Since(begin).Seconds()
+
+		if err == nil {
+			return elapsed, nil
+		}
+
+		if _, ok := err.(recoverableError); !ok {
+			return -1, err
+		}
+
+		if reportFailureFunc != nil {
+			reportFailureFunc()
+		}
+		level.Debug(t.logger).Log("msg", "failed to send batch, retrying", "err", err)
+
+		time.Sleep(time.Duration(backoff))
+		backoff = backoff * 2
+
+		if backoff > t.cfg.MaxBackoff {
+			backoff = t.cfg.MaxBackoff
+		}
+	}
+}
+
+func (t *QueueManager) sendMetadataWithBackoff(ctx context.Context, metadata []prompb.MetricMetadata) error {
+	req, _, err := buildWriteRequest(nil, metadata, nil)
+	if err != nil {
+		return err
+	}
+
+	elapsed, err := t.sendWriteRequestWithBackoff(ctx, req, nil)
+	if elapsed > 0 {
+		t.sentMetadataDuration.Observe(elapsed)
```

Comment:
`0` is valid observation if super quick I think, so let's observe if err == nil?

Reply:
I agree 0 is a valid observation, I'll change that. 

However, I don't feel observing err is the way to go. In the case of a transport error (or not), you still want to know how long the request took.

Reply:
Agree, that's why do it only on err == nil?

Reply:
A transport error will return an error, so it wouldn't be measured.

Reply:
I also think you should measure even in the case of an error. You definitely want to observe things like timeouts.

Reply:
Resolving as we do need to observe in the case of _some_ errors.

Reply:
Ok, just we need to make sure function really calculate elapsed ALWAYS on error - it's unusual for function to care about other returned args if error is retured so might be suprising

Reply:
To summary, the problem here is that we need to observe everything that's not a context cancellation error. I have made that explicit and now return the default when we don't have value as originally suggested.